### PR TITLE
LPS-115593 LPS-115057 Add axis labels, empty state when there are no categories selected

### PR DESCRIPTION
**Pull request by dgarciasarai**
**Description**

The main goal of this pull request is to show the axis names in the chart and put an empty state when the user unchecked all the categories from the legend.

**Changes**

* Use Lexicon colors for the bars
* Display axis names in the chart
* Empty state when there are no categories selected.

**Feature flag: YES**

The Audit Graph is hidden. The configuration file to show the chart is here: https://issues.liferay.com/browse/LPS-115805

**Module**

```
~/liferay-portal/modules/apps/content-dashboard/content-dashboard-web
```

**Screenshots**

Screenshot 2020-06-29 at 12 02 29

Screenshot 2020-06-29 at 13 24 00

cc: @bryceosterhaus @wincent

### CODE REVIEW

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
Hey @wincent, since @bryceosterhaus it's on PTO this week, can you review this PR?
I just rebased it and resolve the conflicts. Thanks.

**Comment by dgarciasarai** [comment]
ci:test:sf

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 1]
```diff
@@ -132,28 +149,51 @@ export default function AuditBarChart({rtl, vocabularies}) {
...
+					</ClayCheckbox>
+				))}
+			</>
+		);
 	};
 
+	const allCheckboxesNotChecked = Object.keys(checkboxes).every(
+		(i) => checkboxes[i] === false
+	);
+
```
[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js:183]: Agreed 👍

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/css/main.scss [hunk 2]
```diff
@@ -26,3 +26,11 @@
 	transform: translateX(-40%);
 	z-index: 1;
 }
+
+.no-categories-selected {
+	top: 18%;
```
[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/css/main.scss:31]: What do you propose? Using `px` instead? 🤔

[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/css/main.scss:31]: Can we use `transform` here? To position horizontally and vertically? 🤔 
```css
transform: translateY(60%) translateX(-40%);
```

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/css/main.scss:31]: > What do you propose? Using `px` instead?

The beauty of `#JustSayin` is that I don't have to propose anything! 😀 

I am wondering whether you can put this in a flex container and use `align-content` to get the placement that you want.

Note: if you can't come up with an "elegant" solution for this, that is ok — it's not a blocker, but as you know, I was `#JustSayin`... 😂

> Can we use `transform` here?

I'd say you can, but if you're still using a percent-based offset you won't really have avoided the magic numbers.

> ```css
> transform: translateY(XX%) translateX(-XX%);
> ```

BTW, you can write `translateX(-XX%) translateY(YY%)` as `translate(-XX%, YY%)`.

[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/css/main.scss:31]: The Recharts chart is rendering as an `<svg>` and doesn't allow to render custom components inside it (or I didn't found it...), so I have to position absolutely via CSS 🤷‍♀️

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js [hunk 3]
```diff
@@ -31,6 +32,22 @@ import {shortenNumber} from '../utils/shortenNumber';
 
 export default function AuditBarChart({rtl, vocabularies}) {
 	const auditBarChartData = useMemo(() => {
+		const axisNames = vocabularies.reduce((_acc, category) => {
```
[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js]: It makes sense. It was weird to me too. Using a `reduce` but not using the accumulator... 😢 
I tried what you suggest and it works fine in all the scenarios (at least the three that I mentioned before in previous PRs). Thanks for the feedback 😄

**Comment by dgarciasarai** [comment]
@wincent I just updated the PR with you feedback 👍 Thanks!

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
@wincent should I change anything else? Please, let me know. Thanks 😄

**Comment by wincent** [comment]
ci:test:relevant

**Comment by dgarciasarai** [comment]
Should we pass `ci:forward` instead? Brian's pull requests display how many `ci:forward` attempts and the time it took and I think it's super useful to the CI maintainers.

**Comment by wincent** [comment]
> Should we pass `ci:forward` instead? Brian's pull requests display how many `ci:forward` attempts and the time it took and I think it's super useful to the CI maintainers.

No, due to the changes the CI team made recently, it is not safe to run `ci:forward` without first seeing `ci:test:relevant` green. ie. old flow:

1. `ci:forward`
2. `ci:test:relevant` and `ci:test:stable` get triggered automatically.
3. If everything green, forwards automatically, unless there are failures in common with upstream in some batches (eg. functional test batches).

That meant that in practice, you would often see tests that are supposedly green and had "no unique failures", but you'd have to manually forward them if the "failures in common with upstream" were functional tests *and* you were sure that the *same* failures really were already present in upstream (ie. it is possible your tests were failing for *different* reasons).

New flow: they changed `ci:forward` to forward even if there are common-with-upstream failures in functional test batches. This means that if you do `ci:forward` *first*, you risk shipping new breakage upstream whenever it overlaps with existing breakage. So what you have to do now, to be safe, is:

1. `ci:test:relevant`.
2. If green, analyze "failures in common with upstream" to be sure that you haven't caused *new* issues to be introduced.
3. `ci:forward` once you are convinced all is good.
4. Forwarding happens.

Does that make sense?

It unfortunately adds latency to every single PR, but it eliminates the number one cause for having to do manual forwards. Arguably, it's a bad system, but less bad than the alternative.

**Comment by dgarciasarai** [comment]
Ok, I didn't have that information. Is there a thread to read about this new flow on Slack (maybe)?

**Comment by wincent** [comment]
> Ok, I didn't have that information. Is there a thread to read about this new flow on Slack (maybe)?

Yes, [here is the announcement](https://liferay.slack.com/archives/C19PWQ74J/p1592420332050100).

The trade off here is subtle. We've reduced the friction of `ci:forward` by reducing its safety guarantees, and we've shifted the burden for guaranteeing the correctness of changes further towards developers (because it requires them to know/understand what `ci:forward` really means and what the subtleties of it are, and it requires them to execute the steps in the right order, at the right time, with the right response to the results). So, this reduces frustration for developers ("why isn't my pull forwarding even though it has no unique failures"), at the cost of potentially reducing stability in the product. This is what people asked for, so that's the tradeoff we got. (The "real" fix is to make functional test batches much more granular, so that we don't get false "common" results, but that is A Hard Problem™ and I wouldn't expect much progress on that front in the immediate future.)

**Comment by wincent** [comment]
FWIW, asked about sluggish CI [here](https://liferay.slack.com/archives/C19PWQ74J/p1593617024078800).

**Comment by wincent** [comment]
From `#ci`:

> The `ci:test:relevant` top-level-build appears to have died due to a bad slave node. Please retry.

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
Screenshot 2020-07-01 -173951-uKKwiHgd@2x

not-that-old

**Comment by dgarciasarai** [comment]
ci:stop

**Pull request closed without merging**